right/src/slave_scheduler.c

```diff
@@ -93,13 +93,16 @@ static void slaveSchedulerCallback(I2C_Type *base, i2c_master_handle_t *handle,
             currentSlave->init(currentSlave->perDriverId);
         }
 
-        status_t currentStatus = currentSlave->update(currentSlave->perDriverId);
+        bool shouldYield = true;
+        status_t currentStatus = currentSlave->update(currentSlave->perDriverId, &shouldYield);
         if (IS_STATUS_I2C_ERROR(currentStatus)) {
             LogI2cError(currentSlaveId, currentStatus);
         }
         isTransferScheduled = currentStatus != kStatus_Uhk_IdleSlave && currentStatus != kStatus_Uhk_IdleCycle;
 
-        if (currentStatus != kStatus_Uhk_IdleCycle) {
```

Comment:
I have removed this condition effectively, since keeping the logic intact would significantly decrease code readability, and I didn't see a reason to keep it.

Reply:
-> Should kboot idle cycles at https://github.com/UltimateHackingKeyboard/firmware/pull/518/files#diff-0f726b9e03a7f7b68ce24085975f37582511def44889dc80637f55b7b930b25aL44 be forced to not yield?

Reply:
`kStatus_Uhk_IdleCycle` is used deliberately, and it shouldn't be ignored. When a slave returns it, the scheduler should not move to the next slave. It's used when no I2C transaction is done in a phase, only validation, to jump to the next phase quickly. Have you tried to use it instead of `yield`? Maybe it's a viable alternative.

Please only use `kStatus_Uhk_IdleCycle`, `yield`, or any similar solution to alter what's really necessary, which I believe is only the touchpad slave driver, to keep this change minimal. My memories about the codebase are a bit foggy, and the smaller the scope of the change, the easier for me to understand it.

Great finding about touchpad timing!

Reply:
> kStatus_Uhk_IdleCycle is used deliberately, and it shouldn't be ignored. When a slave returns it, the scheduler should not move to the next slave. It's used when no I2C transaction is done in a phase, only validation, to jump to the next phase quickly. Have you tried to use it instead of yield? Maybe it's a viable alternative.

Shouldn't I return the I2C status when a phase consists of I2C communication?

Also, am I supposed to return kStatus_Success in order to yield when no I2C communication took place?

> Please only use kStatus_Uhk_IdleCycle, yield, or any similar solution to alter what's really necessary, which I believe is only the touchpad slave driver, to keep this change minimal. My memories about the codebase are a bit foggy, and the smaller the scope of the change, the easier for me to understand it.

Shouldn't we try to reduce the overall response time? 

I mean, if I make touchpad process its messages atomically, but leave the module slave driver intact, the left half/module update will take twice the time of touchpad update.

Reply:
Indeed, when the phase consists of an I2C communication, the status code code should be returned, so `kStatus_Uhk_IdleCycle` is only useful when no communication happens.

I think the most elegant way is to return the status value and the yield value in a struct by slaves.

Instead of the `yield` name, I'd rather use the more intuitive `hold` name to hold control and not allow the scheduler to move to the next slave. I don't understand why 3 states would be useful, and I'd use a boolean `hold`.

When no communication takes place, I'd use a new `kStatus_None` value, and then instead of using `kStatus_Uhk_IdleCycle`, it'd make more sense to return `{.status=kStatus_None, .hold=true}`.

I don't see a way to reduce the overall response time. We can make practical compromises, but we have only so much I2C bandwidth.

Reply:
> I don't understand why 3 states would be useful, and I'd use a boolean `hold`.

Because if you want to have the noop-cycle-skipping scheduling behaviour coupled to `kStatus_Uhk_IdleCycle` (which sounds more reasonable than writing out the `.hold = true` everywhere because it carries additional semantics), then you need to let the slave scheduler know whether it should respect the `hold` or the `kStatus_Uhk_IdleCycle`?

Also, it might be better to explicitly specify whether or not I (as a human being and author of the commit) care about the scheduling policy rather than having to specify the `.hold = true` everywhere, because if I explicitly have to state the value everywere, then I actually erase the information of whether I (as the commit author) do care or do not and why so.

> When no communication takes place, I'd use a new `kStatus_None` value, and then instead of using `kStatus_Uhk_IdleCycle`, it'd make more sense to return `{.status=kStatus_None, .hold=true}`.

What about `kStatus_Uhk_IdleSlave`... ? You are breaking the pattern...

> I don't see a way to reduce the overall response time. We can make practical compromises, but we have only so much I2C bandwidth.

Well: make sure that all modules update at same pace w.r.t. the number of actually completed update cycles (proposed behaviour) rather than w.r.t. completed I2C exchanges (current behaviour)

In order to model the impact, assume:
- each I2c transfer takes 1ms (not realistic)
- module update cycle takes 2 transfers
- touchpad update cycle takes 6 transfers
- assume only 2 uhk modules and 1 touchpad module to be in the update loop
- (I also assume that in module code, data to be sent are prepared at receiving the request rather than when sending them, which means that the message actually sits one slave scheduler cycle without being updated, therefore increasing the actual latency by further factor of 1.5x.)

strategy | current scheduling | make just touchpad processing atomic | atomic touchpad and atomic modules
--|--|--|--
module update rate | 6ms      | 16ms | 10 ms
module latency | 9ms          | 24ms | 10 ms
touchpad update rate | 18ms  | 8ms   | 10 ms

For instance, If I apply the blocking/holding policy to touchpad only (as you seem to suggest), then the update loop will look like:

- left half 1 I2c transfer
- left module 1 I2c transfer
- touchpad 1 update cycle == 6 I2c transfers
- (touchpad update loop completed here)
- left half 1 I2c transfer
- (left half update loop completed here, since it takes 2 I2c transfers)
- left module 1 I2c transfer
- (left module update loop completed here, since it takes 2 I2c transfers)
- touchpad 1 update cycle == 6 I2c transfers
- (touchpad update loop completed here)

Reply:
Thanks for elaborating. I agree that we should cluster module I2C transactions for the modules to have the same refresh rate.

I forgot that there's both `kStatus_Uhk_IdleSlave` and `kStatus_Uhk_IdleCycle`. I'd ditch them both and use `kStatus_None` instead with and also use `.lock = true | false`.

I'd always respect `.hold` regardless of `.status`.

`.hold = true` shouldn't be "explicitly stated everywhere", only at the beginning of slave update functions, which is already done for `status`.

Reply:
> `.lock = true | false`.
What is `.lock`?

> I'd always respect `.hold` regardless of `.status`.
> 
> `.hold = true` shouldn't be "explicitly stated everywhere", only at the beginning of slave update functions, which is already done for `status`.

I mean, that these two statements imply that `.hold = true` has to be manually included everywhere where `kStatus_Uhk_IdleCycle` is returned. Furthermore, the idea of IdleCycle phases being contracted is entirely removed from the code, and is required to be filled in by the user (by either including `.hold = true` or by deliberately omitting it).

This means doing exactly what I meant by removing the `kStatus_Uhk_IdleCycle` condition in the first post.

Reply:
There is no problem. 

It is just that you stated that:

> kStatus_Uhk_IdleCycle is used deliberately, and it shouldn't be ignored. When a slave returns it, the scheduler should not move to the next slave. 

That is in direct contradiction with respecting `.hold`, because respecting `.hold` means that `kStatus_Uhk_IdleCycle` is no longer semantically coupled with the requirement of directly skipping to the next phase of the same slave.

Reply:
Ok, I think we're on the same page then.